test/test_torch.py

```diff
@@ -6261,18 +6261,31 @@ def map_location(storage, loc):
 
         def load_bytes():
             with open(test_file_path, 'rb') as f:
-                data = io.BytesIO(f.read())
-            return data
+                return io.BytesIO(f.read())
 
         fileobject_lambdas = [lambda: test_file_path, load_bytes]
-        map_locations = [map_location, {'cuda:0': 'cpu'}, 'cpu']
+        cpu_map_locations = [map_location, {'cuda:0': 'cpu'}, 'cpu', torch.device('cpu')]
+        gpu_map_locations = [
+            {'cuda:0': 'cuda:0'},
+            'cuda',
+            'cuda:{}'.format(torch.cuda.device_count() - 1),
+            'cuda:0',
+            torch.device('cuda'),
+            torch.device('cuda', 0)
+        ]
 
         for fileobject_lambda in fileobject_lambdas:
-            for map_location in map_locations:
+            for map_location in cpu_map_locations:
                 tensor = torch.load(fileobject_lambda(), map_location=map_location)
                 self.assertIsInstance(tensor, torch.FloatTensor)
                 self.assertEqual(tensor, torch.FloatTensor([[1.0, 2.0], [3.0, 4.0]]))
 
+            if torch.cuda.is_available():
+                for map_location in gpu_map_locations:
+                    tensor = torch.load(fileobject_lambda(), map_location=map_location)
+                    self.assertIsInstance(tensor, torch.cuda.FloatTensor)
+                    self.assertEqual(tensor, torch.cuda.FloatTensor([[1.0, 2.0], [3.0, 4.0]]))
```

Comment:
Can we please also have this:
```python
if isinstance(map_location, str):
    self.assertEqual(tensor.device, torch.device(map_location))
elif isinstance(map_location, torch.device):
    self.assertEqual(tensor.device, map_location)
```

Reply:
We get the error message: 
`device(type='cuda', index=0) != device(type='cuda')` when we do this.

I'm not sure how valid of an error message this is. Should I change Device.cpp so that index=0 is equivalent to no index?

Reply:
Done